lib/linked-list.js

```diff
@@ -19,148 +20,232 @@ class LinkedList {
     method to delete the first node found with specified value
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
     delete(value) {
-        throw new Error('This method has\'t been implemented yet!')
+        (this.#head?.value === value)?(this.#head = this.#head.next):false;
+        let current = this.#head
+        while (current) {
+            (current.next?.value === value)? (current.next = current.next.next):
+            current = current.next
+        }
     }
+    
 
     /*
     method to print all the values in the linked list
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
     visit() {
         const helper_list = []
-        current = this.#head;
-
+        let current = this.#head;
         while (current) {
             helper_list.push(current.value);
             current = current.next;
         }
 
-        console.log(helper_list.toString());
+        console.log(helper_list);
     }
 
 
     /*
     method to reverse the singly linked list
     note: the nodes should be moved and not just the values in the nodes
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
     reverse() {
-        throw new Error("This method hasn't been implemented yet!");
+        let current = this.#head
+        let prev
+        let next
+        while (current) {
+            if (current===this.#head) {
+                prev = current
+                current = current.next
+            } else {
+                next = current.next
+                current.next = prev
+                prev = current
+                current = next
+            }
+        }
+        this.#head = prev
     }
 
     // Advanced Exercises
     /*
     returns the value at the middle element in the singly linked list
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
      */
     findMiddleValue() {
-        throw new Error("This method hasn't been implemented yet!");
+        let len = this.length()
+        const evenCheck = len%2
+        return evenCheck ? this.getAtIndex(Math.floor(len/2)):(this.getAtIndex((len/2)))
+
     }
 
     /*
     find the nth node from the end and return its value
     assume indexing starts at 0 while counting to n
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
     findNthFromEnd(n) {
-        throw new Error("This method hasn't been implemented yet!");
+        return(this.getAtIndex(this.length()-n))
     }
 
     /*
     checks if the linked list has a cycle. A cycle exists if any node in the
     linked list links to a node already visited.
     returns true if a cycle is found, false otherwise.
-    Time Complexity: ?
-    Space Complexity: ?
+    Time Complexity: O(n)
+    Space Complexity: O(1)
     */
     hasCycle() {
```

Comment:
What if the loop isn't back to the `head?  What if it loops back to the 3rd node in the list?